### ISSUE TITLE
Notes on chat fine-tuning and datacontent

### ISSUE THREAD

**Issue by RonanKMcGovern** [jzhang38/TinyLlama#55]
I adapted TimDettmers filtered Openassistant dataset in order for it to take the Llama 2 prompt format (e.g. with INST), see [here](https://huggingface.co/datasets/Trelis/openassistant-llama-style/).

I then fine-tuned TinyLlama (using a full fine-tune of all LoRA modules) at the 1T token checkpoint, see [here](https://huggingface.co/Trelis/TinyLlama-1.1B-intermediate-step-480k-1T-chat-llama-style).

Observations:
A. TinyLlama seems to have issues emitting an EOS (< /s > token). For example:
```
<s> [INST] What planets are in our solar system? [/INST] 1. Mercury

2. Venus

3. Earth

4. Mars

5. Jupiter

6. Saturn

7. Uranus

8. Neptune

9. Pluto

10. Ceres

11. Callisto

12. ...
```

This leads me to wonder are BOS and, particularly, EOS tokens being used in pre-training (e.g. < s > and < /s >)?

B. I notice that when inferencing the raw 1T checkpoint (i.e. not chat fine-tuned), it is common to see ### in the response:
```
<s> [INST] Generate a python code snippet to add two numbers. [/INST] 

### [INST] Generate a python code snippet to add two numbers.

### [INST] Generate a python code snippet to add two numbers.

...
```
I'm somewhat surprised to see this '###'. Does this mean there are some chat fine-tuning or instruct fine-tuning datasets in the pre-training datasets?

**Comment by jzhang38** [comment]
Thanks for the question. 

BOS is used during pretraining to separate documents. On a hindsight, I should have used EOS. 

But I am surprised that the model had trouble ending at EOS,  if you add such tokens in the FT stage. Are you sure yo have calculated the loss on the EOS token? 

While the pretraining data used is strictly Slimpajama and Starcodercode, I cannot answer if texts similar to chat or ### exists in those datasets.

**Comment by RonanKMcGovern** [comment]
Ahhh, yes, that may have been important - the inclusion of EOS. Is there a way to include it going forward?

I've run follow up fine tuning on TinyLlama and Llama 7B and can confirm that 7B fine-tunes well:
```
<s> [INST] What planets are in our solar system? [/INST] 1. The Sun. 2. Mercury. 3. Venus. 4. Earth. 5. Mars. 6. Jupiter. 7. Saturn. 8. Uranus. 9. Neptune.</s>



<s> [INST] What are the first five numbers in the Fibonacci series? [/INST] 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578, 5702887, 92



<s> [INST] Generate a python code snippet to add two numbers. [/INST] 
python
def add_two_numbers(x, y):
   return x + y


</s>

```

By contrast, TinyLlama does not emit </s>, which makes sense to me if it was not trained as such:
```
<s> [INST] What planets are in our solar system? [/INST] 1. Mercury 2. Venus 3. Earth 4. Mars 5. Jupiter 6. Saturn 7. Uranus 8. Neptune 9. Pluto 10. Ceres 11. Ceres 12. Ceres 13. Ceres 14. Ceres 15. Ceres 16. Ceres 17. Ceres 18. Ceres 19. Ceres 20. Ceres 21. Ceres 22. Ceres 23. Ceres 24. Ceres 25. Ceres 26. Ceres 27. Ceres 28. Ceres 29. Ceres 30. Ceres 31. Ceres 32. Ceres 33. Ceres 34. Ceres 35. Ceres



<s> [INST] What are the first five numbers in the Fibonacci series? [/INST] 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 888, 1397, 2196, 3495, 5594, 8993, 14492, 23391, 37780, 61073, 88862, 139751, 219642, 349533, 559424, 899325, 1449216, 2339107, 3778008, 6107309,



<s> [INST] Generate a python code snippet to add two numbers. [/INST] 
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
...
#
#
#
#
```

I can try to fine-tune TinyLlama for longer, but it's looking difficult to get the </s> to emit.

**Comment by RonanKMcGovern** [comment]
@jzhang38 I ran another fine-tune today, this time using a custom trainer where I know that the EOS token is included in attention and the loss calculation.

The results were the same - i.e. TinyLlama does not emit an EOS token.

My sense from here is that chatml is then the best/only way to go as that explicitly defines a new token for BOS and EOS.

**Comment by ZeguanXiao** [comment]
@RonanKMcGovern Are you using a full fine-tuning or a LoRA fine-tuning?

**Comment by RonanKMcGovern** [comment]
I was doing LoRa but probably I wasn't additionally training the embedding tokens - which is what I would recommend now if one needs to change the behaviour of tokens (Which is what is needed here).
